nodedb.go

```diff
@@ -444,34 +444,53 @@ func (ndb *nodeDB) DeleteVersionsFrom(version int64) error {
 
 	// First, delete all active nodes in the current (latest) version whose node version is after
 	// the given version.
-	err = ndb.deleteNodesFrom(version, root)
-	if err != nil {
-		return err
-	}
-
-	// Next, delete orphans:
-	// - Delete orphan entries *and referred nodes* with fromVersion >= version
-	// - Delete orphan entries with toVersion >= version-1 (since orphans at latest are not orphans)
-	err = ndb.traverseOrphans(func(key, hash []byte) error {
-		var fromVersion, toVersion int64
-		orphanKeyFormat.Scan(key, &toVersion, &fromVersion)
-
-		if fromVersion >= version {
-			if err = ndb.batch.Delete(key); err != nil {
-				return err
-			}
-			if err = ndb.batch.Delete(ndb.nodeKey(hash)); err != nil {
-				return err
+	if !fastMode {
+		err = ndb.deleteNodesFrom(version, root)
+		if err != nil {
+			return err
+		}
+		// Next, delete orphans:
+		// - Delete orphan entries *and referred nodes* with fromVersion >= version
+		// - Delete orphan entries with toVersion >= version-1 (since orphans at latest are not orphans)
+		err = ndb.traverseOrphans(func(key, hash []byte) error {
+			var fromVersion, toVersion int64
+			orphanKeyFormat.Scan(key, &toVersion, &fromVersion)
+
+			if fromVersion >= version {
+				if err = ndb.batch.Delete(key); err != nil {
+					return err
+				}
+				if err = ndb.batch.Delete(ndb.nodeKey(hash)); err != nil {
+					return err
+				}
+				ndb.nodeCache.Remove(hash)
+			} else if toVersion >= version-1 {
+				if err = ndb.batch.Delete(key); err != nil {
+					return err
+				}
 			}
-			ndb.nodeCache.Remove(hash)
-		} else if toVersion >= version-1 {
-			if err = ndb.batch.Delete(key); err != nil {
-				return err
+			return nil
+		})
+	} else {
+		err = ndb.traverseOrphansVersion(version-1, func(key, hash []byte) error {
```

Comment:
why do we change a version here?

Reply:
since `toVersion` in orphan records is current `version-1`

Reply:
thanks, let's add a comment in the code.